### PR TITLE
chore: Update TransactionController to v13 and ApprovalController to v3.5.2

### PR DESCRIPTION
**Pull request by tommasini**
## **Description**
* Update patches of approval controller and Transaction controller accordingly to the core branch
* Created a script for the approval controller and updated the script for transaction controller patch

Transaction Controller v9 breaking changes:
* Remove fetchAll method :white_check_mark: (Already addressed on the past)
* Remove prepareUnsignedEthTx and getCommonConfiguration methods :white_check_mark: (no changes needed)
* Update properties of RemoteTransactionSourceRequest type  :white_check_mark: (no changes needed)
* Move all but first argument to options bag in addTransaction method :white_check_mark: (Already addressed on the past)
* Add isSupportedNetwork method to RemoteTransactionSource interface :white_check_mark: (no changes needed)
* Add required getSelectedAddress callback argument to constructor :white_check_mark:  (Already addressed on the past)
 
Transaction Controller v10 breaking changes:
* Rename rawTransaction to rawTx in the transaction metadata
rawTransaction  could be found on the first layer of [TransactionMetadata](https://github.com/MetaMask/core/pull/1624/files#diff-50efcf23d8b038bbbd28fabd611a837999ac8939eddacdefc31d5ad77db22e84R84) (I can't identify this on the store data, will tag the confirmations team to review this migration) 🧪

TransactionController v11 breaking changes:
* Rename the transactionHash property to hash in the transaction metadata  🧪
* Add disableSendFlowHistory constructor option :white_check_mark: (added on the current work)
* Add disableHistory constructor option :white_check_mark: (added on the current work)
* Rename the transaction object to txParams in the transaction metadata  🧪
* Remove incomingTransactions.apiKey constructor option :white_check_mark: (added on the current work)
* Remove unused FetchAllOptions type from TransactionController  :white_check_mark: (No changes needed)
* Remove apiKey property from RemoteTransactionSourceRequest type (I can't identify but it seems that there is no changes needed on our side)

TransactionController v12 breaking changes:
* Change TransactionMeta.chainId to be required :white_check_mark: (no changes needed)
* Use only chainId to determine if a transaction belongs to the current network :white_check_mark: (no changes needed)

TransactionController v13 breaking changes:
* Add required getCurrentNetworkEIP1559Compatibility callback arguments to constructor :white_check_mark:  (added on the current work)

 🧪 - It was tested while debugging but better re-test when every bug is fixed
 
 
 QUESTION FOR CONFIRMATIONS TEAM:
* Make sure SubmitHistoryEntry didn't rename rawTransaction property to rawTx. (History is disabled [here](https://github.com/MetaMask/metamask-mobile/pull/9088/files#diff-1a656ef42f595bb9c6c1251b29808fd70d31da7a3e1a6e9be85574ad866dc1b0R1110))
* Make sure that readableValue property is correctly used on this [line](https://github.com/MetaMask/metamask-mobile/blob/80ce3177c0f394d0306f2cfb783781f188e2a9c9/app/components/Nav/Main/RootRPCMethodsUI.js#L317)  since it doesn't exist on [TransactionParams](https://github.com/MetaMask/core/blob/12bd75d0038d6bbb3b22d0cb523bf9d7a3bcc8b0/packages/transaction-controller/src/types.ts#L246)
* multiLayerL1FeeTotal property is correctly used on this [line](https://github.com/MetaMask/metamask-mobile/blob/e571db750d1fda8e3e21000fa1190a51836d1810/app/components/UI/TransactionElement/TransactionDetails/index.js#L180) since since it doesn't exist on [TransactionParams](https://github.com/MetaMask/core/blob/12bd75d0038d6bbb3b22d0cb523bf9d7a3bcc8b0/packages/transaction-controller/src/types.ts#L246)?

## **Related issues**

Fixes: https://github.com/MetaMask/mobile-planning/issues/1627

## **Manual testing steps**

QA build for testing: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/0ce45ce4-6a8a-4501-a9e5-97cf835df703

Perform transaction regression

## **Screenshots/Recordings**

https://docs.google.com/document/d/1ctuiK-J8M_n-h_HWRde7Odyttyo3VlLWI5D9tn4YCI8/edit

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: patches/@metamask+transaction-controller+13.0.0.patch [hunk 1]
```diff
@@ -0,0 +1,1649 @@
+diff --git a/node_modules/@metamask/transaction-controller/dist/.patch.txt b/node_modules/@metamask/transaction-controller/dist/.patch.txt
+new file mode 100644
+index 0000000..550de56
+--- /dev/null
++++ b/node_modules/@metamask/transaction-controller/dist/.patch.txt
+@@ -0,0 +1,7 @@
++PATCH GENERATED FROM MetaMask/core branch: patch/mobile-transaction-controller-11-0-0
```
[the remark this answers is not in the document]
Comment (vinistevam) [reply patches/@metamask+transaction-controller+13.0.0.patch]: Thanks! I've updated the patch and pushed the changes.

File: app/components/Views/confirmations/Approval/index.js [hunk 2]
```diff
@@ -453,7 +454,16 @@ class Approval extends PureComponent {
       );
 
       const fullTx = transactions.find(({ id }) => id === transaction.id);
-      const updatedTx = { ...fullTx, transaction };
+
+      const updatedTx = {
```
Comment (vinistevam) [inline app/components/Views/confirmations/Approval/index.js:458]: For transparency and audit purposes, this adjustment is essential until a refactor is completed on the reducer.

File: app/components/Views/confirmations/ApproveView/Approve/index.js [hunk 3]
```diff
@@ -540,7 +540,15 @@ class Approve extends PureComponent {
       );
 
       const fullTx = transactions.find(({ id }) => id === transaction.id);
-      const updatedTx = { ...fullTx, transaction };
+
+      const updatedTx = {
```
Comment (vinistevam) [inline app/components/Views/confirmations/ApproveView/Approve/index.js:544]: Same here

File: app/components/Views/confirmations/Approval/index.js (hunk 2 above)
Comment (tommasini) [reply app/components/Views/confirmations/Approval/index.js:458]: Yeah! That refactor will happen on the update from v^16 to v^24 since on v^18 we migrate the base controller to v^4 on Transaction Controller

This is the [ticket](https://github.com/MetaMask/mobile-planning/issues/1700) where this refactor will be done

File: app/components/Views/confirmations/ApproveView/Approve/index.js (hunk 3 above)
Comment (tommasini) [reply app/components/Views/confirmations/ApproveView/Approve/index.js:544]: https://github.com/MetaMask/metamask-mobile/pull/9088#discussion_r1572177137

**Comment by mcmire** [comment]
Nice work!

File: app/components/Views/confirmations/SendFlow/Confirm/index.js [hunk 4]
```diff
@@ -826,8 +826,6 @@ class Confirm extends PureComponent {
         ExtendedKeyringTypes.ledger,
       ]);
 
-      await this.persistTransactionParameters(transaction);
```
Comment (matthewwalsh0) [inline app/components/Views/confirmations/SendFlow/Confirm/index.js:829]: Why was this removed?

I recall adding this to fix an issue where the gas fees were not persisted in the send flow.

File: app/core/Engine.ts [hunk 5]
```diff
@@ -1104,11 +1104,12 @@ class Engine {
         onNetworkStateChange: (listener) =>
           this.controllerMessenger.subscribe(
             AppConstants.NETWORK_STATE_CHANGE_EVENT,
-            //@ts-expect-error TransactionController needs to be updated to v13 for this error disappears
             listener,
           ),
         // @ts-expect-error at this point in time the provider will be defined by the `networkController.initializeProvider`
         provider: networkController.getProviderAndBlockTracker().provider,
+        disableSendFlowHistory: true,
```
Comment (matthewwalsh0) [inline app/core/Engine.ts:1111]: For the sake of maintenance, should we keep these in alphabetical order?

File: app/core/Engine.ts [hunk 6]
```diff
@@ -1075,12 +1075,12 @@ class Engine {
...
           networkController.getProviderAndBlockTracker().blockTracker,
         getGasFeeEstimates: () => gasFeeController.fetchGasFeeEstimates(),
-        //@ts-expect-error TransactionController needs to be updated to v13 for this error disappears
+        getCurrentNetworkEIP1559Compatibility:
+          networkController.getEIP1559Compatibility.bind(networkController),
         getNetworkState: () => networkController.state,
         getSelectedAddress: () =>
           accountsController.getSelectedAccount().address,
         incomingTransactions: {
-          apiKey: process.env.MM_ETHERSCAN_KEY,
```
Comment (matthewwalsh0) [inline app/core/Engine.ts:1083]: ❤️

File: app/core/NotificationManager.js [hunk 7]
```diff
@@ -331,20 +331,24 @@ class NotificationManager {
   watchSubmittedTransaction(transaction, speedUp = false) {
     if (transaction.silent) return false;
     const { TransactionController } = Engine.context;
-    const nonce = transaction.transaction.nonce;
+    const transactionMeta = TransactionController.state.transactions.find(
```
Comment (matthewwalsh0) [inline app/core/NotificationManager.js:334]: Why are we searching the state for the metadata? Is the `transaction` argument not the metadata we need?

File: app/store/migrations/039.test.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        engine: {
+          backgroundState: { TransactionController: null },
+        },
+      }),
+      errorMessage: "Migration 39: Invalid TransactionController state: 'null'",
+      scenario: 'transactionController is invalid',
+    },
+  ];
+  for (const { errorMessage, scenario, state } of invalidStates) {
+    it(`should capture exception if ${scenario}`, async () => {
```
Comment (matthewwalsh0) [inline app/store/migrations/039.test.ts:97]: Minor, but would it be more idiomatic to use `it.each` rather than a loop to generate these tests?

File: app/store/migrations/039.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        `Migration 39: Invalid TransactionController state: '${state.engine.backgroundState.TransactionController}'`,
+      ),
+    );
+    return state;
+  }
+
+  const transactionControllerState =
+    state.engine.backgroundState.TransactionController;
+
+  if (!hasProperty(transactionControllerState, 'transactions')) {
```
Comment (matthewwalsh0) [inline app/store/migrations/039.ts:43]: Do we need this granularity in errors?

Could we just have one check such as:

`if(!isObject(state?.engine?.backgroundState?.TransactionController?.transactions))`

File: app/store/migrations/039.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    captureException(
+      new Error(
+        `Migration 39: Missing transactions property from TransactionController: '${typeof state
+          .engine.backgroundState.TransactionController}'`,
+      ),
+    );
+    return state;
+  }
+
+  if ((transactionControllerState.transactions as []).length) {
```
Comment (matthewwalsh0) [inline app/store/migrations/039.ts:53]: If we're just iterating over the array, then do we gain anything by checking the length first?

The result should be the same if the array is empty.

Comment (matthewwalsh0) [inline app/store/migrations/039.ts:53]: Rather than casting twice and referencing the full path multiple times, should we just define it one with optional chaining and then do our check on it, and then iterate over it?

File: scripts/patch-approval-controller.sh
Comment (matthewwalsh0) [inline scripts/patch-approval-controller.sh:1]: Not needed in this PR, but given the overlap here, should we just create a single more generic `patch-controller.sh` that we invoke from both scripts in `package.json`?

**Review by matthewwalsh0 — COMMENTED** [review]
Epic and excellent PR @tommasini, thanks!

File: app/core/Engine.ts (hunk 5 above)
Comment (tommasini) [reply app/core/Engine.ts:1111]: Amazing suggestion! Will change this!

File: app/store/migrations/039.test.ts (hunk 8 above)
Comment (tommasini) [reply app/store/migrations/039.test.ts:97]: Hmm, never thought about it! Will try it! Thanks!

File: app/store/migrations/039.ts (hunk 9 above)
Comment (tommasini) [reply app/store/migrations/039.ts:43]: I'm not sure this would work, just because `transactions` property is an array and not an object by definition

File: scripts/patch-approval-controller.sh
Comment (tommasini) [reply scripts/patch-approval-controller.sh:1]: That would probably be a very good idea for every package we have patches, I believe that will help a lot on the migration to ESM when jumping the core major version to v125! 

Genius! Thanks Matthew! Will give more thoughts to this on next upgrades and think on a solution!

File: app/components/Views/confirmations/SendFlow/Confirm/index.js (hunk 4 above)
Comment (tommasini) [reply app/components/Views/confirmations/SendFlow/Confirm/index.js:829]: Ah already remembered. this was removed because it is duplicated, it is persisting on line 823 of this file

File: app/store/migrations/039.ts (hunk 9 above)
Comment (matthewwalsh0) [reply app/store/migrations/039.ts:43]: Good point, but what I meant was we could have just one error check on the specific state we need. We could do `Array.isArray` instead for example but just once.

File: app/core/NotificationManager.js (hunk 7 above)
Comment (tommasini) [reply app/core/NotificationManager.js:334]: No :( at that point in time, the nonce is not present in the transaction argument, but it is on the transactions object of the TransactionController 

I believe the redux reducer of the transaction on the application in general needs a refactor, but this PR was getting big so we talked about addressing it on next ones!

Let me know your thoughts about it

**Comment by christopherferreira9** [comment]
All good on SDK. Also checked general deeplinking behaviour around transactions and the same for WalletConnect.

**Comment by seaona** [comment]
hi there :wave: leaving QA report - will finish updating it today @tommasini 

## ERC20
### :green_circle:  Autodetect, Send, Edit
(from inside wallet)

https://github.com/MetaMask/metamask-mobile/assets/54408225/e82cec6f-0e3f-489c-9f28-2ed457c23c3f

### :green_circle:  Import, Send, Approve - with different token decimals

(from dapp)

https://github.com/MetaMask/metamask-mobile/assets/54408225/f6577aa1-ef3e-45b7-ae32-2b7aa592cc53

### :green_circle:  Send, Approve with Deeplinks

https://github.com/MetaMask/metamask-mobile/assets/54408225/8d6963fb-c707-44ae-b1d5-b1b65a1e14b7

### Send, Approve with Wallet Connect

https://github.com/MetaMask/metamask-mobile/assets/54408225/2d079d8a-1d47-4db0-8b28-a6401bf972fc

## ERC721
### Send, Approve, SetApprovalForAll

(from dapp)

https://github.com/MetaMask/metamask-mobile/assets/54408225/e4e56cef-a5ea-4081-9714-6dfb42fea9dd

(from inside wallet)

https://github.com/MetaMask/metamask-mobile/assets/54408225/c38b4eaf-c779-4d09-9dec-eef90f73d2d8


## Migrations
from7.21 to feature branch
(pending vids)

File: app/components/Views/confirmations/SendFlow/Confirm/index.js (hunk 4 above)
Comment (tommasini) [reply app/components/Views/confirmations/SendFlow/Confirm/index.js:829]: More precise this line: https://github.com/MetaMask/metamask-mobile/blob/5e16d729abc42602a04ef1a2129ebe9555749d26/app/components/Views/confirmations/SendFlow/Confirm/index.js#L823

File: app/store/migrations/039.ts (hunk 9 above)
Comment (tommasini) [reply app/store/migrations/039.ts:43]: Actually that's it's an amazing suggestion!!! Yes That would let this cleaner

File: app/core/Engine.ts (hunk 5 above)
Comment (tommasini) [reply app/core/Engine.ts]: Addressed on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9088/commits/e32b95fc090ea0f1433ffa23c196c65699c798ff)

File: app/store/migrations/039.ts (hunk 9 above)
Comment (tommasini) [reply app/store/migrations/039.ts]: Addressed on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9088/commits/e32b95fc090ea0f1433ffa23c196c65699c798ff)

File: app/store/migrations/039.ts (hunk 10 above)
Comment (tommasini) [reply app/store/migrations/039.ts]: I believe we don't need since we are using the Array.isArray, on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9088/commits/e32b95fc090ea0f1433ffa23c196c65699c798ff)

**Review by jpuri — DISMISSED** [review]
LGTM

**Review by OGPoyraz — DISMISSED** [review]
Amazing job Tomas 🚀

File: app/store/migrations/039.ts (hunk 10 above)
Comment (tommasini) [reply app/store/migrations/039.ts]: Yess, addressed on this commit : https://github.com/MetaMask/metamask-mobile/pull/9088/commits/e32b95fc090ea0f1433ffa23c196c65699c798ff

File: app/store/migrations/039.test.ts (hunk 8 above)
Comment (tommasini) [reply app/store/migrations/039.test.ts]: Implemented on this commit: https://github.com/MetaMask/metamask-mobile/pull/9088/commits/e32b95fc090ea0f1433ffa23c196c65699c798ff

**Review by legobeat — DISMISSED** [review]
LGTM!

**Comment by seaona** [comment]
## QA Findings
(in progress)

### Send ETH displayed as Contract Deployment

I think this issue might be only happening in this branch

https://github.com/MetaMask/metamask-mobile/issues/9433

File: app/store/migrations/040.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
+import { isObject } from '@metamask/utils';
+import { captureException } from '@sentry/react-native';
+
+export default function migrate(state: unknown) {
+  if (!isObject(state)) {
```
Comment (Cal-L) [inline app/store/migrations/040.ts:5]: Use previously created utility function to perform initial checks

**Review by Cal-L — COMMENTED** [review]
Left a comment

File: app/store/migrations/040.ts (hunk 11 above)
Comment (tommasini) [reply app/store/migrations/040.ts]: Thanks for the reminder Cal!!! Addressed on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9088/commits/dfdd382757dcf9b3f53d8acafbdc04ccae470e43)

**Review by Cal-L — COMMENTED** [review]
Migrations code looks good from a platform perspective. Will defer to the transactions code to confirmations team to review.

**Review by vinistevam — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

**Comment by seaona** [comment]
thank you for the recording uploads @tommasini . I think the PR looks good from QA seeing the issues have been fixed
I shared the PR with @sleepytanya and @pnarayanaswamy in the case something else is needed

**Review by dbrans — APPROVED** [review]

**Review by legobeat — APPROVED** [review]
LGTM!

**Review by sethkfman — APPROVED** [review]
LGTM
